lom/python/src/common/engine_rpc_if.py

```diff
@@ -0,0 +1,290 @@
+# ActionResponseData. For shutdown, ResData is empty
+#
+@dataclass
+class MsgSendServerResponse:
+    ResType:        int
+    ResData:        {}
+
+
+
+# Response sent as ReqData for ReqType = TypeNotifyActionHeartbeat
+@dataclass
+class MsgNotifyHeartbeat:
+    Action:     str
+    Timestamp:  int
+
+
+# MsgSendServerResponse:ResData for ResType == TypeServerRequestAction
+@dataclass
+class ActionResponseData:
+    Action:             str
+    InstanceId:         str
+    AnomalyInstanceId:  str
+    AnomalyKey:         str
+    Response:           str
+    ResultCode:         int
+    ResultStr:          str
+
+
+# Response sent by engine is of the type LoMResponse
+# RespData is empty for all requests except for TypeRecvServerRequest
+# For request with ReqType = TypeRecvServerRequest, RespData carries
+# ServerRequestData.
+@dataclass
+class LoMResponse:
+    ResultCode:         int
+    ResultStr:          str
+    RespData:           {}
+
+
+# Request to plugin/action or Plugin Mgr is sent as response to
+# client's request with ReqType = TypeRecvServerRequest
+# ReqType can be TypeServerRequestAction or TypeServerRequestShutdown
+# For ReqType == TypeServerRequestAction, the ReqData is set to
+# ActionRequestData. For shutdown, ReqData is empty
+#
+@dataclass
+class ServerRequestData:
+    ReqType:        int
+    ReqData:        {}
+
+
+# ActionRequestData carries the detailed request to action/plugin.
+#
+@dataclass
+class ActionRequestData:
+    Action:             str
+    InstanceId:         str
+    AnomalyInstanceId:  str
+    AnomalyKey:         str
+    Timeout:            int
+    Context:            [ActionResponseData]
+
+
+# Client registration is once per session. Both connection & name are cached
+# and reused until de-reg / re-reg client.
+#
+clientName = None
+clientConn = None
+
+def connectServer() -> bool:
+    global clientConn
+
+    # Register client *always* connect. Close if any existing.
+    disconnectServer()
+
+    try:
+        conn = socket.create_connection(("localhost", LOM_RPC_JSON_PORT))
+        clientConn = conn
+    except socket.gaierror as e: 
+        log_error("Address-related error connecting to server: %s" % e)
+    except socket.error as e:
+        log_error("Connection error: %s" % e)
+    return clientConn != None
+
+
+def disconnectServer():
+    global clientConn
+
+    if clientConn != None:
+        clientConn.close()
+        clientConn = None
+
+
+def retFailedResponse(req: LoMRequest, code: int, msg: str) -> LoMResponse:
+    msg = "Failed: Code={} msg={} req={}".format(code, msg, req)
+    log_error(msg)
+    return LoMResponse(code, msg, {})
+
+
+ReqId = 0
+
+# Send LoMRequest as JSON string
+# Recv LoMResponse as JSON string
+# Load parses string as SimpleNamespace with members matching LoMResponse
+#
+def sendAndReceive(req: LoMRequest) -> SimpleNamespace:
+    # Just a uniq ID for RPC message to check against response
```

Comment:
can we use GUID ?

Reply:
This is just to match req & res in this connection. We are synchronous and send only one req. So any number would do, as no other response is expected. Even if you are sending many, just a sequence will do, as it just needs to be unique across outstanding requests.